src/parameter_server.cpp

```diff
@@ -4,51 +4,63 @@
 #include <Utils.h>
 #include <config.h>
 
+#include <S3.h>
 #include <stdlib.h>
 #include <cstdint>
 #include <string>
 
 #include <S3.h>
 #include <gflags/gflags.h>
 
+#include "Utils.h"
+
 DEFINE_int64(nworkers, -1, "number of workers");
 DEFINE_int64(rank, -1, "rank");
 DEFINE_string(config, "", "config");
-DEFINE_string(ps_ip, PS_IP, "parameter server ip");
-DEFINE_int64(ps_port, PS_PORT, "parameter server port");
+
+DEFINE_string(ps_ip, PS_IP, "parameter server ips comma separated");
+DEFINE_string(ps_port, "1337", "parameter server ports comma separated");
 
 static const uint64_t GB = (1024*1024*1024);
 static const uint32_t SIZE = 1;
 
-void run_tasks(int rank, int nworkers,
-    int batch_size, const cirrus::Configuration& config,
-    const std::string& ps_ip,
-    uint64_t ps_port) {
-
+void run_tasks(int rank,
+               int nworkers,
+               int batch_size,
+               const cirrus::Configuration& config,
+               std::vector<std::string> ps_ips,
+               std::vector<uint64_t> ps_ports) {
   std::cout << "Run tasks rank: " << rank << std::endl;
   int features_per_sample = config.get_num_features();
   int samples_per_batch = config.get_minibatch_size();
 
   if (rank == PS_SPARSE_SERVER_TASK_RANK) {
-    cirrus::PSSparseServerTask st((1 << config.get_model_bits()) + 1,
-        batch_size, samples_per_batch, features_per_sample,
-        nworkers, rank, ps_ip, ps_port);
+    cirrus::PSSparseServerTask st(
+        (1 << config.get_model_bits()) + 1, batch_size, samples_per_batch,
+        features_per_sample, nworkers, rank, ps_ips[0], ps_ports[0]);
     st.run(config);
   } else if (rank >= WORKERS_BASE && rank < WORKERS_BASE + nworkers) {
     /**
      * Worker tasks run here
      * Number of tasks is determined by the value of nworkers
      */
     if (config.get_model_type() == cirrus::Configuration::LOGISTICREGRESSION) {
-      cirrus::LogisticSparseTaskS3 lt(features_per_sample,
-          batch_size, samples_per_batch, features_per_sample,
-          nworkers, rank, ps_ip, ps_port);
-      lt.run(config, rank - WORKERS_BASE);
+      if (ps_ips.size() > 1) {
+        cirrus::LogisticSparseTaskS3 lt(features_per_sample, batch_size,
+                                        samples_per_batch, features_per_sample,
+                                        nworkers, rank, ps_ips, ps_ports);
+        lt.run(config, rank - WORKERS_BASE);
+      } else {
+        cirrus::LogisticSparseTaskS3 lt(features_per_sample, batch_size,
+                                        samples_per_batch, features_per_sample,
+                                        nworkers, rank, ps_ips[0], ps_ports[0]);
```

Comment:
Here and below: we should just pass the ps_ips regardless of how many there are.

Reply:
Fixed